src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -81,73 +118,93 @@ AsyncLogWriter::AsyncLogWriter()
   : _flush_sem(0), _lock(), _data_available(false),
     _initialized(false),
     _stats() {
+
+  size_t page_size = os::vm_page_size();
+  size_t size = align_up(AsyncLogBufferSize / 2, page_size);
+
+  char* buf0 = static_cast<char* >(os::malloc(size, mtLogging));
```

Comment:
Is there a particular reason that AsyncLogWriter allocates and handles the resources intended for `Buffer` instead of `Buffer` doing it itself in its ctr and dtr?

Reply:
I have no idea how to handle the error that malloc returns NULL in ctor. Maybe I am paranoid. glibc/malloc rarely fails because it maps but not commit a memory segment.  We also define the constraint of AsyncLogBufferSize in globals.hpp.

Reply:
You're right, I forgot how impossible that case is.  Being paranoid with this is fine, imho.

Reply:
Probably extremely rare, since argument parsing needs C-heap too, for malloc=NULL to hit right here. Not impossible though (limits, sbrk exhausted, etc). Note that we have other libcs too.

But I agree with @jdksjolen original remark. Buffer managing its memory would be more straightforward. No need to handle malloc errors if you just use NEW_C_HEAP_ARRAY and let it fatal out in case of an OOM. Its what we do all over hotspot. Would also save you the ugly cast (why do C++ casts have to be so verbose..).

Reply:
okay. thanks. I will update this.

I prefer to detect this allocation failure and fall back to synchronous logging. This is for resource-constrained java. '-Xlog:async' shouldn't contribute a runtime error.